metadata/src/main/java/org/apache/kafka/controller/BrokerHeartbeatManager.java

```diff
@@ -223,6 +283,21 @@ public BrokerHeartbeatState next() {
         }
     }
 
+    /**
+     * The maximum number of timed out heartbeats to count.
+     */
+    static final int DEFAULT_TIMED_OUT_HEARTBEAT_COUNT_MAX = 1000;
+
+    /**
+     * The time period over which to track timed out heartbeats.
+     */
+    static final long DEFAULT_TIMED_OUT_HEARTBEAT_COUNT_WINDOW_NS = TimeUnit.MINUTES.toNanos(5);
+
+    /**
+     * The number of heartbeats to notice missing before we go into overload.
+     */
+    static final int DEFAULT_TIMED_OUT_HEARTBEAT_OVERLOAD_THRESHOLD = 3;
+
```

Comment:
Whatever we set this to (and currently we always take this default value) will determine how long it would take for a controller to leave the overloaded state, right?  So if we see a series of missed heartbeats that puts us into the overloaded state, and then we don't see any more missed heartbeats, it would take this amount of time to leave that overloaded state?

Consider the following case.  We are overloaded.  And then some broker crashes.  How long would we want it to take before we move leadership away for any partitions led by that suddenly-dead broker?  Ideally the session timeout, of course, but that won't happen if we are overloaded -- will it take 5 minutes as currently coded? That seems like a long time.

I think the only thing we can say is that we should move leadership away as soon as we have a correct view of the cluster.  The question then becomes: how do we know when we have good information?

A fixed amount of time like this with no missed heartbeats -- that seems too simple.  We likely have a reasonable view of the cluster if we see enough heartbeats without missing any such that it is likely that all brokers have had a chance to communicate with us.  So it seems like the broker heartbeat interval should factor into it somehow.

Let's say the broker heartbeat interval is 2 seconds.  Right now we consider a heartbeat missed if it is more than half the heartbeat interval (i.e. 1 second) old.  Let's say there are 60 brokers in the cluster.  If we see something on the order of 60 heartbeats in row without having missing any, then we probably have a reasonable view.

So maybe it isn't having missed no more than N heartbeats in some fixed time window.  Maybe we define "not overloaded" as having seen some string of heartbeats uninterrupted without seeing a missed one?  So every time we miss a heartbeat we go into the "overloaded" state and we reset a counter of contiguous successfully processed heartbeats to 0.  Whenever we see a heartbeat and process it in time we increase that counter.  We are overloaded when the `contiguousHeartbeatsProcessedInTime` counter is equal to or exceeds the broker count.

I'm not sure I ave it right, but I think this is a discussion worth pursuing.

Reply:
Thinking about this some more, I believe the key requirement is to accurately understand when we do not have a correct view of the cluster.  This allows us to handle  2 important cases: not fencing a broker if its session times out but we think we could have missed enough heartbeats to make the decision to fence the wrong decision; and fencing a broker if its session times out whenever we think we have accurate information.  I think I addressed the second part (detect and fence a crashed broker as quick as possible) in the comment above: I believe we have accurate information if we see a contiguous series of N successfully-processed heartbeats with no intervening timed-out heartbeats where N is perhaps the cluster size.  For the first part (don't fence if we think the decision to do so could be wrong) assume the broker session is 18 seconds and the heartbeat interval is 2 seconds.  That means we would need to miss 9 heartbeats for a broker in order to incorrectly fence it.  Maybe we keep track of the last time we had enough contiguous successful heartbeats (which, if we aren't missing any would always be very recent).  But then as soon as we miss one we increment the missed count and reset the contiguous count to 0.  When we successfully process a heartbeat we increment the contiguous count and, if it reaches the necessary threshold N (which is on the order of the cluster size) we reset the missed count to 0.  We can fence brokers only when the missed count approaches the session/interval ratio (i.e. 18/2 = 9 by default).

We can tweak this to be a bit more conservative.  Maybe we need N (the number of contiguous heartbeats seen to assure us we have good visibility) to be 1.5 or 2 times the broker count instead of just the broker count.  Maybe the missed count only has to exceed half the session/interval ratio (so only missing 5 heartbeats without seeing N successfully-processed ones in a row would prevent fencing instead of 9 by default).

WDYT?

Reply:
I agree with you that 5 minutes is way too long to "mute" broker removals. That was a miss in the orignial PR. (As a side note, broker session is not 18 seconds. It is 9 seconds.)

I thought about this more and I think we may not need the overload state I originally wanted to introduce at all. We can simply do some basic processing on the heartbeat when we time it out. Specifically, we can update the "last seen time" of the broker. This will avoid the "congestion meltdown" behavior without introducing a new state.

Reply:
> As a side note, broker session is not 18 seconds. It is 9 seconds.

I was confused by the existence of this, which sets 18 seconds: https://github.com/apache/kafka/blob/d557f7997359daa5853d446d9093995577540de2/metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java#L77

But the above always get overridden by the value of `broker.session.timeout.ms` in the config, which does default to 9 seconds (https://github.com/apache/kafka/blob/d557f7997359daa5853d446d9093995577540de2/core/src/main/scala/kafka/server/KafkaConfig.scala#L83).

So yes, it is 9 seconds as you state above.  I wonder if the default in `ClusterControlManager.java` should be changed to match the config's default.

Reply:
I'll update it